### DISCUSSION TITLE
Conditional dynamic recipients

### DISCUSSION THREAD

**Discussion by croxton** [solspace/craft-freeform#898] [Questions & Answers] [answered]
I'm trying to replicate the conditional notifications feature of WordPress’ Gravity Forms plugin. Specifically, I have a country dropdown and a department dropdown. The recipients for the admin notification email may depend on which country is selected AND/OR which department is selected. More than one notification may be sent.

E.g.
Country: USA
Department: HR
=> Send to hr@usa.company.com

Country: USA
Department: Customer services
=> Send to customer@usa.company.com

Country: France or Germany
Department: HR
=> Send to hr@eu.company.com

Country: France or Germany
Department: Customer services
=> Send to customer@eu.company.com

Country: USA
Department: [any selected]
=> Send to admin@usa.company.com

(etc)

Is this best handled with some custom logic in a webhook, or could it be done with conditional fields? Or with javascript - assuming all possible recipients were listed for a Dynamic Recipients field, would freeform allow the value of that field to be set with js? What would you suggest as the best approach?

**Comment by kjmartens** [comment, marked as answer]
What you would likely have to do here is have a field for `Country`, and then have a `Department` field for each country. For example, `department_usa`, `department_france`, etc. Then populate these Department fields with the relevant department names and email addresses, and show/hide the Department fields based on the selected Country.

It's not ideal, but it's how it would need to be done if using Freeform's conditional logic. The other option would be custom JS approach. 🙂

**Comment by croxton** [comment]
That would get pretty hard to manage in the freeform interface given there are several hundred countries! 

It also looks like a `Dynamic Recipients` field doesn't allow one email address to be used in multiple options, so a JS solution is probably out too, as there is some crossover unfortunately.

I guess it'll have to be a webhook that sends it's own email(s) based on the form selects.

**Reply by kjmartens** [reply]
> That would get pretty hard to manage in the freeform interface given there are several hundred countries!

Sorry, I wasn't sure if you just had a few for the site (e.g. USA, Canada, and a few European ones, etc). Definitely would but ugly to attempt this approach for every country. 🥴 

> It also looks like a `Dynamic Recipients` field doesn't allow one email address to be used in multiple options, so a JS solution is probably out too, as there is some crossover unfortunately.

This is correct as well. For what it's worth, this is finally addressed in Freeform 5. 🙂

**Comment by croxton** [comment]
For future reference, I found a way to do this in Freeform 4 using the `EVENT_BEFORE_SEND` event and a `formNotifications` Section with entries representing each notification. The entry has an `email` field, a `subject` field for the subject and a `countries` matrix with one block containing a `country` field. The custom notification logic is triggered when a form has the Admin Recipient set to a special email address, e.g. `custom-notifications@my-website.com`.

<img width="996" alt="Screenshot 2024-02-20 at 18 14 27" src="https://github.com/solspace/craft-freeform/assets/279127/e6b5834e-c460-4755-89ff-344856a92236">


In the `init()` method of a custom module/plugin (e.g. `\modules\custommodule`):

```php
Event::on(
    MailerService::class,
    MailerService::EVENT_BEFORE_SEND,
    new \modules\custommodule\handlers\FormNotifications()
);

```

The handler looks like this:

```php
<?php
namespace modules\custommodule\handlers;

use Craft;
use Solspace\Freeform\Events\Mailer\SendEmailEvent;
use craft\elements\Entry;

class FormNotifications {

    public string $customNotificationEmail = 'custom-notifications@my-website.com';

    /**
     * @throws \JsonException
     */
    public function __invoke(SendEmailEvent $event)
    {
        $message = $event->getMessage();
        $formValues = $event->getFieldValues();

        // get the first admin email address
        $to = '';
        $toArray = $message->getTo();
        if (!empty($toArray)) {
            $to = array_key_first($toArray);
        }

        // Special admin email is used to trigger custom notifications handling
        if ($to === $this->customNotificationEmail) {

            $sendTo = [];

            // Get all form notification entries
            $entries = Entry::find()
                ->section('formNotifications')
                ->limit(null)
                ->with(['countries'])
                ->all();

            // Selected country.
            // Convert selected country to corresponding 2-letter code
            // if the full country name was used for the value
            $selectedCountry = null;
            if (isset($formValues['country'])) {
                $selectedCountry = $formValues['country']->getValue();
            }

            if ($selectedCountry && !preg_match("/^[A-Z]{2}$/", $selectedCountry)) {
                try {
                    $countriesData = json_decode(file_get_contents(Craft::$app->getVendorPath() . '/solspace/craft-freeform/packages/plugin/src/Library/Factories/Data/countries.json'), true, 512, JSON_THROW_ON_ERROR);
                    $countriesData = array_flip($countriesData);
                    if (array_key_exists($selectedCountry, $countriesData)) {
                        $selectedCountry = $countriesData[$selectedCountry];
                    }
                } catch (\JsonException $exception) {
                    // do nothing
                }
            }

            // Subject field
            $selectedSubject = null;
            if (isset($formValues['subject'])) {
                $selectedSubject = $formValues['subject']->getValue();
            }

            // Process notification entries
            foreach ($entries as $entry) {
                $sendEmail = true;

                // get field values from notification entry
                $fields = $entry->getFieldValues(['countries', 'email', 'subject']);

                // The email to send the notification to
                $email = $fields['email']; // required field

                // Limit to countries, if specified in the notification entry
                $countries = [];
                foreach ($fields['countries'] as $country) {
                    $countries[] = (string) $country->country;
                }
                if ($selectedCountry && !empty($countries)) {
                    $sendEmail = false;
                    if (in_array($selectedCountry, $countries, true)) {
                        $sendEmail = true;
                    }
                }

                // Limit to subject(s), if specified in the notification entry
                $subject = $fields['subject'] ?? '';
                if ($sendEmail && $selectedSubject && !empty($subject)) {
                    $subjects = array_map('trim', explode(',', $subject));
                    if (!empty($subjects)) {
                        $sendEmail = false;
                        if (in_array($selectedSubject, $subjects, true)) {
                            $sendEmail = true;
                        }
                    }
                }

                // Add the notification email
                if ($sendEmail) {
                    $sendTo[] = $email;
                }
            }

            if (!empty($sendTo)) {
                // overwrite the sending email addresses
                $message->setTo($sendTo);
            } else {
                // no notifications to send
                $event->isValid = false;
            }
        }
    }
}
```
